internal/service/flexcluster/state_transition_test.go

```diff
@@ -0,0 +1,120 @@
+package flexcluster_test
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/service/flexcluster"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"go.mongodb.org/atlas-sdk/v20240805004/admin"
+	"go.mongodb.org/atlas-sdk/v20240805004/mockadmin"
+)
+
+var (
+	IdleState      = "IDLE"
+	CreatingState  = "CREATING"
+	UpdatingState  = "UPDATING"
+	DeletingState  = "DELETING"
+	RepairingState = "REPAIRING"
+	sc500          = conversion.IntPtr(500)
+	sc200          = conversion.IntPtr(200)
+	sc404          = conversion.IntPtr(404)
+	clusterName    = "clusterName"
+	requestParams  = &admin.GetFlexClusterApiParams{
+		GroupId: "groupId",
+		Name:    clusterName,
+	}
+)
+
+type testCase struct {
+	expectedState *string
+	name          string
+	mockResponses []response
+	desiredStates []string
+	pendingStates []string
+	expectedError bool
+}
+
+func TestFlexClusterStateTransition(t *testing.T) {
+	testCases := []testCase{
+		{
+			name: "Successful transition to IDLE",
+			mockResponses: []response{
+				{state: &CreatingState, statusCode: sc200},
+				{state: &IdleState, statusCode: sc200},
+			},
+			expectedState: &IdleState,
+			expectedError: false,
+			desiredStates: []string{IdleState},
+			pendingStates: []string{CreatingState},
+		},
+		{
+			name: "Error when API returns 5XX",
+			mockResponses: []response{
+				{statusCode: sc500, err: errors.New("Internal server error")},
+			},
+			expectedState: nil,
+			expectedError: true,
+			desiredStates: []string{IdleState},
+			pendingStates: []string{CreatingState},
+		},
+		{
+			name: "Deleted state when API returns 404",
+			mockResponses: []response{
+				{state: &DeletingState, statusCode: sc200},
+				{statusCode: sc404, err: errors.New("Not found")},
+			},
+			expectedState: nil,
+			expectedError: true,
+			desiredStates: []string{IdleState},
+			pendingStates: []string{UpdatingState, DeletingState},
```

Comment:
To keep in mind, I believe delete operation is only using `DeletingState` as a pending state. Might be good to make small functions that encapsulate the pending and target states ([search deployment reference](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/searchdeployment/state_transition_search_deployment.go#L18)) so we ensure that in testing the same states are being used.